daemon/albatrossd.ml

```diff
@@ -145,12 +145,23 @@ let jump _ influx tmpdir dbdir =
   | Error (`Msg msg) -> Logs.err (fun m -> m "bailing out: %s" msg)
   | Ok old_unikernels ->
     Lwt_main.run
-      (let unix_connect s =
-         Vmm_lwt.connect Lwt_unix.PF_UNIX (Lwt_unix.ADDR_UNIX (socket_path s))
+      (let rec unix_connect ~retries s =
+         Vmm_lwt.connect Lwt_unix.PF_UNIX (Lwt_unix.ADDR_UNIX (socket_path s)) >>= fun x ->
+         (match x with
+         | Some x -> Lwt.return (Some x)
+         | None -> if (retries-1 != 0) then begin
+                     Logs.err (fun m -> m "unable to connect to %s, retrying in 5 seconds" (socket_path s));
+                     Lwt_unix.sleep 5.0 >>= fun () ->
+                     unix_connect ~retries:(retries-1) s
+                   end else
+                     Lwt.return_none)
```

Comment:
I'm not too convinced of this approach. At the moment, albatrossd requires the console and log daemon to be executed before albatrossd is started. Once either of them disappears (and albatrossd notices), albatrossd gives up and terminates.

With this change, albatrossd could idle not starting up due to issues in the other daemons. Currently, the order is enforced by the service scripts + rc scripts.

OTOH, it'd be nice to be able to start and stop and restart the daemons completely independently (esp. log and stats -- console may be a bit tricky (if it's reading from any fifos, it is not clear to me whether it can be safely stop (close() the fifo) and started again (open() all fifos)).

Reply:
Yes, if console or log fails or are slow to start `albatrossd` would wait and report which socket it was waiting for -- currently indefinitely, but I guess there could be a `--retries=` with a lower default if you want it to fail. I just noticed that `albatrossd` also opens its own socket first, which isn't ideal as it's not functional yet (but it looks like that could be moved). I don't mind closing this if you'd prefer to keep the solution as it is for now though -- I mostly wrote this to make my own testing easier :)

Reply:
After sleeping over it, I'd be fine with:
- opening the server socket only after the other connections are up and running (well spotted, thanks)
- accepting an optional `retry-connections` command-line argument to use for the connect-on-startup logic. if it is 0, no retry should be done. its default should be some small number (maybe 3)
- an `--enabled-stats` command-line argument (your `--disable-stats`, but inverted) to control whether a connection to the stats is attempted (if `--enable-stats` is used, and the connection to the stats socket fails, albatrossd should terminate)

Reply:
Ok, sounds good -- I'll update the PR with these changes (hopefully later today)